### PR TITLE
Integration Test Optimizations

### LINKED ISSUE #9498: Integration Test Optimizations
### Housekeeping

- [X] I am a maintainer of dbt-core

### Short description

Cache static objects between tests to accelerate integration runs.

### Impact to Other Teams

None.

### Will backports be required?

No.

### Context

_No response_

### PR DESCRIPTION
**Pull request by peterallenwebb**
resolves #9498 

### Problem
 
Integration tests perform a lot of identical rework for every test.

### Solution

Cache important static objects between test runs.

### Checklist

- [x] I have read [the contributing guide](https://github.com/dbt-labs/dbt-core/blob/main/CONTRIBUTING.md) and understand what's expected of me  
- [x] I have run this code in development and it appears to resolve the stated issue  
- [x] This PR includes tests, or tests are not required/relevant for this PR
- [x] This PR has no interface changes (e.g. macros, cli, logs, json artifacts, config files, adapter interface, etc) or this PR has already received feedback and approval from Product or DX
- [x] This PR includes [type annotations](https://docs.python.org/3/library/typing.html) for new and modified functions

### CODE REVIEW

**Review by gshank — APPROVED** [review]

**Review by colin-rogers-dbt — CHANGES_REQUESTED** [review]
Blocking on validation this doesn't impact adapter expectations around macros/test fixtures: https://github.com/dbt-labs/dbt-snowflake/actions/runs/7733189834/job/21084750889

File: core/dbt/tests/fixtures/project.py [hunk 1]
```diff
@@ -517,6 +518,7 @@ def project(
     # Logbook warnings are ignored so we don't have to fork logbook to support python 3.10.
     # This _only_ works for tests in `tests/` that use the project fixture.
     warnings.filterwarnings("ignore", category=DeprecationWarning, module="logbook")
+    enable_test_caching()
```
Comment (colin-rogers-dbt) [inline core/dbt/tests/fixtures/project.py:521]: can we add a way to disable this if we want to?

**Comment by colin-rogers-dbt** [comment]
does this result in test runs caching macros in ways that dbt-core wouldn't do typically? That seems to create a divergence between actual and test behavior of dbt-core

**Comment by colin-rogers-dbt** [comment]
> Blocking on validation this doesn't impact adapter expectations around macros/test fixtures: https://github.com/dbt-labs/dbt-snowflake/actions/runs/7733189834/job/21084750889

Looks like it succeeded so it's probably good to go, will check a couple other adapters

**Comment by peterallenwebb** [comment]
@colin-rogers-dbt I considered that, and wouldn't be doing this if it didn't save us a ton of testing time. With that said, the objects we are caching appear to be static after construction, and are computed as a pure function of their construction parameters. So while I can not completely rule out future complications without deeper knowledge of jinja internals, my sense is that complications are unlikely.

**Review by colin-rogers-dbt — APPROVED** [review]

**Pull request merged**